File: ServiceBinder/app/src/main/java/com/alex/servicebinder/MainActivity.java
package com.alex.servicebinder;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.ServiceConnection;
import android.os.IBinder;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

  private static final String TAG = MainActivity.class.getSimpleName();

  private LocalService service;
  private boolean bound = false;
  private Toast toast;

  /**
   * Defines callbacks for service binding, passed to bindService()
   */
  private ServiceConnection connection = new ServiceConnection() {

    @Override
    public void onServiceConnected(ComponentName className, IBinder service) {
      // We've bound to LocalService, cast the IBinder and get LocalService instance
      LocalService.LocalBinder binder = (LocalService.LocalBinder) service;
      MainActivity.this.service = binder.getService();
      bound = true;
    }

    @Override
    public void onServiceDisconnected(ComponentName arg0) {
      bound = false;
    }
  };

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.activity_main);
  }

  @Override
  protected void onStart() {
    super.onStart();
    // Bind to LocalService
    Intent intent = new Intent(this, LocalService.class);
    bindService(intent, connection, Context.BIND_AUTO_CREATE);
  }

  @Override
  protected void onStop() {
    super.onStop();
    unbindService(connection);
    bound = false;
  }

  public void onButtonClick(View view) {
    Log.i(TAG, "bound = " + bound);
    if (bound) {
      // Call a method from the LocalService.
      // However, if this call were something that might hang, then this request should
      // occur in a separate thread to avoid slowing down the activity performance.
      int num = service.getRandomNumber();
      if (toast != null) {
        toast.cancel();
      }
      toast = Toast.makeText(this, "number: " + num, Toast.LENGTH_SHORT);
      toast.show();
    }
  }
}

File: Services/BoundService/app/src/main/java/ru/abaryzhikov/boundservice/MainActivity.kt
package ru.abaryzhikov.boundservice

import android.content.ComponentName
import android.content.Context
import android.content.Intent
import android.content.ServiceConnection
import android.os.Bundle
import android.os.IBinder
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

@Suppress("ConstantConditionIf")
class MainActivity : AppCompatActivity(), ServiceConnection {

    private var service: BoundService? = null
    private val bound: Boolean
        get() = service != null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        updateTimestampButton.setOnClickListener {
            if (bound) {
                timestampView.text = service!!.timestamp
            }
        }

        bindButton.setOnClickListener {
            val intent = Intent(applicationContext, BoundService::class.java)
            if (EXPLICIT_START) {
                startService(intent)
            }
            bindService(intent, this, Context.BIND_AUTO_CREATE)
        }

        stopButton.setOnClickListener {
            if (bound) {
                unbindService(this)
                service = null
            }
            if (EXPLICIT_START) {
                val intent = Intent(applicationContext, BoundService::class.java)
                stopService(intent)
            }
        }
    }

    override fun onStop() {
        super.onStop()
        if (bound) {
            unbindService(this)
            service = null
        }
    }

    override fun onServiceConnected(name: ComponentName?, binder: IBinder?) {
        Log.v(TAG, "onServiceConnected() called")
        val serviceBinder = binder as? BoundService.ServiceBinder
        service = serviceBinder?.service
    }

    override fun onServiceDisconnected(name: ComponentName?) {
        Log.v(TAG, "onServiceDisconnected() called")
        service = null
    }

    companion object {
        private const val TAG = "MainActivity"
        private const val EXPLICIT_START = false
    }
}
